misc/python/materialize/checks/upsert_shrink_grow.py

```diff
@@ -0,0 +1,124 @@
+        pads = self.pads()
+        return Testdrive(
+            schemas()
+            + dedent(
+                f"""
+                $ kafka-create-topic topic=upsert-update-{name}
+
+                $ kafka-ingest format=avro key-format=avro topic=upsert-update-{name} key-schema=${{keyschema}} schema=${{schema}} repeat=10000
+                {{"key1": "${{kafka-ingest.iteration}}"}} {{"f1": "A${{kafka-ingest.iteration}}{pads[0]}A"}}
+
+                > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${{testdrive.kafka-addr}}';
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn FOR CONFLUENT SCHEMA REGISTRY URL '${{testdrive.schema-registry-url}}';
+
+                > CREATE SOURCE upsert_update_{name}
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-upsert-update-{name}-${{testdrive.seed}}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE UPSERT
+
+                > CREATE MATERIALIZED VIEW upsert_update_{name}_view AS
+                  SELECT LEFT(f1, 1), RIGHT(f1, 1),
+                  COUNT(*) AS c1, COUNT(DISTINCT key1) AS c2, COUNT(DISTINCT f1) AS c3,
+                  MIN(LENGTH(f1)) AS l1, MAX(LENGTH(f1)) AS l2
+                  FROM upsert_update_{name}
+                  GROUP BY LEFT(f1, 1), RIGHT(f1, 1);
+                """
+            )
+        )
+
+    def manipulate(self) -> List[Testdrive]:
+        name = self.name()
+        pads = self.pads()
+        return [
+            Testdrive(schemas() + dedent(s))
+            for s in [
+                f"""
+                $ kafka-ingest format=avro key-format=avro topic=upsert-update-{name} key-schema=${{keyschema}} schema=${{schema}} repeat=10000
+                {{"key1": "${{kafka-ingest.iteration}}"}} {{"f1": "B${{kafka-ingest.iteration}}{pads[1]}B"}}
+                """,
+                f"""
+                $ kafka-ingest format=avro key-format=avro topic=upsert-update-{name} key-schema=${{keyschema}} schema=${{schema}} repeat=10000
+                {{"key1": "${{kafka-ingest.iteration}}"}} {{"f1": "C${{kafka-ingest.iteration}}{pads[2]}C"}}
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        name = self.name()
+        last_pad_length = len(self.pads()[-1])
+        return Testdrive(
+            dedent(
+                f"""
+                > SELECT * FROM upsert_update_{name}_view;
+                C C 10000 10000 10000 {last_pad_length+3} {last_pad_length+6}
+                """
+            )
+        )
+
+    def name(self) -> str:
+        assert False
+
+    def pads(self) -> List[str]:
+        assert False
+
+
+class UpsertUpdateShrink(Check, ShrinkGrow):
+    """Upserts where the data length shrinks"""
+
+    def name(self) -> str:
+        return "shrink"
+
+    def pads(self) -> List[str]:
+        return ["x" * 1024, "x" * 512, "x" * 256]
+
+    def initialize(self) -> Testdrive:
```

Comment:
I needed this cruft to get multiple inheritance to work ... @def- if you can suggest a more concise way that would be much appreciated.

Reply:
First I had a hacky solution:
```diff
diff --git a/misc/python/materialize/checks/all_checks.py b/misc/python/materialize/checks/all_checks.py
index 65195f58c9..b9dcb4a5b2 100644
--- a/misc/python/materialize/checks/all_checks.py
+++ b/misc/python/materialize/checks/all_checks.py
@@ -61,5 +61,6 @@ from materialize.checks.top_k import *  # noqa: F401 F403
 from materialize.checks.update import *  # noqa: F401 F403
 from materialize.checks.upsert import *  # noqa: F401 F403
 from materialize.checks.upsert_shrink_grow import *  # noqa: F401 F403
+del globals()["ShrinkGrow"]
 from materialize.checks.uuid import *  # noqa: F401 F403
 from materialize.checks.window_functions import *  # noqa: F401 F403
diff --git a/misc/python/materialize/checks/upsert_shrink_grow.py b/misc/python/materialize/checks/upsert_shrink_grow.py
index 44756507c9..71c3fc0625 100644
--- a/misc/python/materialize/checks/upsert_shrink_grow.py
+++ b/misc/python/materialize/checks/upsert_shrink_grow.py
@@ -18,7 +18,7 @@ def schemas() -> str:
     return dedent(KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD)


-class ShrinkGrow:
+class ShrinkGrow(Check):
     def initialize(self) -> Testdrive:
         name = self.name()
         pads = self.pads()
@@ -86,7 +86,7 @@ class ShrinkGrow:
         assert False


-class UpsertUpdateShrink(Check, ShrinkGrow):
+class UpsertUpdateShrink(ShrinkGrow):
     """Upserts where the data length shrinks"""

     def name(self) -> str:
@@ -95,17 +95,8 @@ class UpsertUpdateShrink(Check, ShrinkGrow):
     def pads(self) -> List[str]:
         return ["x" * 1024, "x" * 512, "x" * 256]

-    def initialize(self) -> Testdrive:
-        return ShrinkGrow.initialize(self)
-
-    def manipulate(self) -> List[Testdrive]:
-        return ShrinkGrow.manipulate(self)
-
-    def validate(self) -> Testdrive:
-        return ShrinkGrow.validate(self)
-

-class UpsertUpdateGrow(Check, ShrinkGrow):
+class UpsertUpdateGrow(ShrinkGrow):
     """Upserts where the data lenth grows"""

     def name(self) -> str:
@@ -113,12 +104,3 @@ class UpsertUpdateGrow(Check, ShrinkGrow):

     def pads(self) -> List[str]:
         return ["x" * 256, "x" * 512, "x" * 1024]
-
-    def initialize(self) -> Testdrive:
-        return ShrinkGrow.initialize(self)
-
-    def manipulate(self) -> List[Testdrive]:
-        return ShrinkGrow.manipulate(self)
-
-    def validate(self) -> Testdrive:
-        return ShrinkGrow.validate(self)
```
Then I realized that the order of inherited classes matters. Works like this:
```diff
diff --git a/misc/python/materialize/checks/upsert_shrink_grow.py b/misc/python/materialize/checks/upsert_shrink_grow.py
index 44756507c9..c05442557b 100644
--- a/misc/python/materialize/checks/upsert_shrink_grow.py
+++ b/misc/python/materialize/checks/upsert_shrink_grow.py
@@ -86,7 +86,7 @@ class ShrinkGrow:
         assert False


-class UpsertUpdateShrink(Check, ShrinkGrow):
+class UpsertUpdateShrink(ShrinkGrow, Check):
     """Upserts where the data length shrinks"""

     def name(self) -> str:
@@ -95,17 +95,8 @@ class UpsertUpdateShrink(Check, ShrinkGrow):
     def pads(self) -> List[str]:
         return ["x" * 1024, "x" * 512, "x" * 256]

-    def initialize(self) -> Testdrive:
-        return ShrinkGrow.initialize(self)
-
-    def manipulate(self) -> List[Testdrive]:
-        return ShrinkGrow.manipulate(self)
-
-    def validate(self) -> Testdrive:
-        return ShrinkGrow.validate(self)
-

-class UpsertUpdateGrow(Check, ShrinkGrow):
+class UpsertUpdateGrow(ShrinkGrow, Check):
     """Upserts where the data lenth grows"""

     def name(self) -> str:
@@ -113,12 +104,3 @@ class UpsertUpdateGrow(Check, ShrinkGrow):

     def pads(self) -> List[str]:
         return ["x" * 256, "x" * 512, "x" * 1024]
-
-    def initialize(self) -> Testdrive:
-        return ShrinkGrow.initialize(self)
-
-    def manipulate(self) -> List[Testdrive]:
-        return ShrinkGrow.manipulate(self)
-
-    def validate(self) -> Testdrive:
-        return ShrinkGrow.validate(self)
```
I just pushed it, hope that's ok.

Reply:
Yes, thanks, squashed.